Migrating to Font-Awesome 5

NishealJ opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Font-Awesome provides vector icons and social icons to oppia, we are currently using Font-Awesome 4 and will be migrating to FontAwesome 5. This Migration will update all the current icons in the codebase and will provide speed improvements.

I've briefed about the approach to this migration [here](https://docs.google.com/document/d/1yBqfSNPwAl3aURxG05CZiosSMbj6jZHts6avcWfDM4A/edit?usp=sharing). Also read the comments  [here](https://github.com/oppia/oppia/pull/6614).

For the testing doc visit [here](https://docs.google.com/document/d/1woPFXlUhv-0uDR0ck4XGNId0IkjjdDeidvBbykxAelE/edit)


## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

NishealJ committed:
Initial Upgrade

NishealJ committed:
Fix build.py for webfont dir

NishealJ committed:
Second upgrade

NishealJ committed:
More fa-icons changed

NishealJ committed:
update app.yaml

apb7 commented:
Hi @NishealJ, the lint tests are failing due to the file limit extension:
```
Running third-party size check
------------------------------------------------------
    Number of files in third-party folder: 7363
    ERROR: The third-party folder size exceeded the 7000 files limit.
------------------------------------------------------
``` 

I think you can modify this check to not consider the `skip_files` in app.yaml.

Also, I see this testing doc [here](https://docs.google.com/document/d/1woPFXlUhv-0uDR0ck4XGNId0IkjjdDeidvBbykxAelE/edit) and just wanted to confirm if you have tested these changes on your end. Thanks!

NishealJ commented:
> I think you can modify this check to not consider the `skip_files` in app.yaml.
>
Hi @apb7,  Yes, we need to update the check to consider skip_files.
> Also, I see this testing doc [here](https://docs.google.com/document/d/1woPFXlUhv-0uDR0ck4XGNId0IkjjdDeidvBbykxAelE/edit) and just wanted to confirm if you have tested these changes on your end. Thanks!
>
I missed this in the testing doc I'll update the check and the testing doc Thanks!

NishealJ committed:
Merge remote-tracking branch 'upstream/develop' into font-awe-m

NishealJ committed:
Fix lint

NishealJ committed:
Fix spaces

NishealJ commented:
Hi @apb7,
I've completely tested this migration for the below points and I've updated the same in the testing doc
1. All the `fa-icons` are up-to-date.
2. The needed files of font-awesome are generated and combined successfully with other lib files in `third_party/generated`.
3. Font files needed for font-awesome is copied to the `webfonts` and retrieved from the same folder.
4. Unwanted files of font-awesome libraries are not deployed in GAE.
5. Font-awesome upgrade doesn't affect `md-icons`.
6. Where-ever needed the icons are made regular/bold. 
7. Additionally added css to the `fa-icons` are not altered.
8. `fa` has been completely removed from the codebase and replaced with needed `far` & `fas` is updated. 

PTAL, Thanks!

vojtechjelinek commented:
@NishealJ Hi, could you please add `<link rel="preload" href="…" as="font">` into *base.html* for the font files that will be loaded from the Font Awesome css?

NishealJ committed:
Pre-load fonts

NishealJ committed:
Handel Cache validation

apb7 commented:
Hi @NishealJ, the e2e tests are failing. PTAL!

vojtechjelinek commented on core/templates/dev/head/pages/base.html:
```diff
@@ -62,6 +62,8 @@
 
     <link rel="apple-touch-icon" href="/assets/images/logo/favicon.png">
+    <link rel="preload" href="/third_party/generated/webfonts/fa-solid-900.woff2" as="font" type="font/woff2">
+    <link rel="preload" href="/third_party/generated/webfonts/fa-brands-400.woff2" as="font" type="font/woff2">
```
Please move these after line 37, also don't we use regular sized icons too and thus should preload them?

NishealJ replied:
Hi @apb7 and @vojtechjelinek,
`fa-solid-900.woff2` and `fa-brands-400.woff2 `are widely requested in the codebase that's why I've preloaded only them in base.html.

Also, e2e tests are falling with this error seems like while pre-loading it is unable to locate the hashed font files
`Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/build/third_party/generated/webfonts/fa-solid-900.a7140145ebaaf5fb14e40430af5d25c4.woff2 - Failed to load resource: the server responded with a status of 404 (Not Found)', timestamp: 1559024798692, type: '' }) to equal undefined.`

vojtechjelinek replied:
If you can't resolve this you can remove the preloads and I will add them later myself. I don't want to block you with stuff that's out of the scope of this PR.

NishealJ replied:
yes, i've reverted the changes

vojtechjelinek reviewed (changes requested):
@NishealJ Thanks, one comment!

NishealJ committed:
Add webfonts

NishealJ committed:
change pos preload webfonts

NishealJ commented on core/templates/dev/head/pages/base.html:
```diff
@@ -37,6 +37,8 @@
   <head>
     {% block prerender %}
     {% endblock prerender %}
+    <link rel="preload" href="/third_party/generated/webfonts/fa-solid-900.woff2" as="font" type="font/woff2">
```
Hi @vojtechjelinek, better we keep it after pre-render?

vojtechjelinek replied:
Sure

NishealJ committed:
Temp commit for  testing

NishealJ committed:
Revert temp changse and Add more specific paths

NishealJ committed:
Revert Changes

NishealJ commented on scripts/build.py:
```diff
@@ -103,7 +103,9 @@
     '*.bundle.js',
-    '*.bundle.js.map')
+    '*.bundle.js.map'
+    '*.woff',
```
Hi @vojtechjelinek, yes seems like this is blocking.
I wonder why it still hashes the font-files after adding at-most specific details in `FILEPATHS_NOT_TO_RENAME`.
did i miss something?

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

apb7 commented:
Hi @NishealJ, any progress here? Thanks!

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
change color

NishealJ commented:
Hi @apb7,
I've Resolved the merge conflicts and your comments in the testing doc PTAL Thanks

apb7 reviewed (commented):
Codewise, this PR LGTM. I have left a couple of comments on the testing doc. Once we've tested this thoroughly, I think this can go in. Thanks, @NishealJ!

NishealJ committed:
Fix merge

NishealJ commented:
Hi @apb7, I've resolved your comments in the testing doc and i've tested this thoroughly. Thanks!

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ committed:
Merge remote-tracking branch 'origin/develop' into font-awe-m

NishealJ commented:
Hi @seanlip @vojtechjelinek & @apb7 PTAL Thanks!

seanlip commented on scripts/build.py:
```diff
@@ -47,6 +47,7 @@
     'css', 'third_party.min.css')
 
 FONTS_RELATIVE_DIRECTORY_PATH = os.path.join('fonts', '')
+WEBFONTS_RELATIVE_DIRECTORY_PATH = os.path.join('webfonts', '')
```
I think you need to add more comments throughout this file to explain the difference between "fonts" and "webfonts" dirs.

NishealJ replied:
Hi @seanlip  Added comments PTAL Thanks!

seanlip commented on scripts/build.py:
```diff
@@ -538,6 +542,11 @@ def build_third_party_libs(third_party_directory_path):
         _generate_copy_tasks_for_fonts(
             dependency_filepaths['fonts'], FONTS_DIR))
 
+    ensure_directory_exists(WEBFONTS_DIR)
```
It is confusing that both these dirs (fonts and webfonts) are being copied to dependency_filepaths['fonts']. Can you elaborate more via comments on what's going on here? Are we planning to get rid of the FONTS_DIR in the future (and if so, is there an issue filed for that that clearly describes what remaining work is needed)?

NishealJ replied:
Hi @seanlip, Yes we will be removing the `fonts` directory in the next milestone.

seanlip replied:
OK, file an issue for this please and link to it here.

Don't just write "Remove the fonts directory" btw -- the issue description should clearly describe all the different steps that will be done and how we will test that the removal is safe. Write it so that a hypothetical new contributor can take it up if possible. Thanks!

NishealJ replied:
Hi @seanlip created an issue #6988 Thanks!

seanlip reviewed (changes requested):
Thanks @NishealJ, a couple comments.

I think maybe @vojtechjelinek should take a look at this. @vojtechjelinek -- is the approach taken here (for webfonts) ok?

seanlip commented:
@NishealJ -- there is no link to a changelog in the testing doc?

seanlip commented:
I also left some notes in the testing doc to request some changes to the visual look-and-feel of the icons -- some of them are now too big.

NishealJ committed:
Changes size of icons

NishealJ commented:
> I also left some notes in the testing doc to request some changes to the visual look-and-feel of the icons -- some of them are now too big.

Hi @seanlip, I'ved addressed your comments Thanks!

NishealJ commented:
> @NishealJ -- there is no link to a changelog in the testing doc?

Added Now Thanks!

seanlip commented:
(Just FYI, I think you haven't finished replying to all review comments yet, so I'll wait to review until you've addressed all review comments.)

NishealJ committed:
add comments

NishealJ commented:
> (Just FYI, I think you haven't finished replying to all review comments yet, so I'll wait to review until you've addressed all review comments.)

Hi @seanlip, replied to all the review comments

seanlip commented on scripts/build.py:
```diff
@@ -47,6 +47,10 @@
     'css', 'third_party.min.css')
 
 FONTS_RELATIVE_DIRECTORY_PATH = os.path.join('fonts', '')
+# /webfonts is need for font-awesome-5, due to changes
```
need --> needed

NishealJ replied:
Done

seanlip commented on scripts/build.py:
```diff
@@ -47,6 +47,10 @@
 
 FONTS_RELATIVE_DIRECTORY_PATH = os.path.join('fonts', '')
+# /webfonts is need for font-awesome-5, due to changes
+# in its directory structure hence, FAM-5 now dont use /fonts
```
dont --> doesn't

NishealJ replied:
Done

seanlip commented on scripts/build.py:
```diff
@@ -47,6 +47,10 @@
 FONTS_RELATIVE_DIRECTORY_PATH = os.path.join('fonts', '')
+# /webfonts is need for font-awesome-5, due to changes
+# in its directory structure hence, FAM-5 now dont use /fonts
+# but /fonts is need by BootStrap and will be removed later
```
need --> needed

NishealJ replied:
Done

seanlip reviewed (approved):
Looks good! Just a few docstring changes and it's good to go.

@vojtechjelinek PTAL?

seanlip commented:
Assigning @aks681 @ankita240796 @kevinlee12 @nithusha21 @vojtechjelinek @apb7 as codeowners.

@NishealJ please resolve remaining comments + merge conflicts. Thanks!

NishealJ committed:
merge conflicts

NishealJ committed:
update comments

NishealJ commented:
> Assigning @aks681 @ankita240796 @kevinlee12 @nithusha21 @vojtechjelinek @apb7 as codeowners.
> 
> @NishealJ please resolve remaining comments + merge conflicts. Thanks!

Hi @seanlip, Done the needed changes and resolved the merge conflicts Thanks!

ankita240796 reviewed (approved):
LGTM from codeowner's perspective, Thanks @NishealJ!

kevinlee12 reviewed (approved):
lgtm!

apb7 reviewed (approved):
LGTM, thanks, @NishealJ!

NishealJ commented:
Hi @vojtechjelinek & @nithusha21, PTAL Thanks!

nithusha21 reviewed (approved):
LGTM as a codeowner.

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!
